pkg/controller/controller.go

```diff
@@ -86,6 +86,10 @@ type Controller interface {
 
 	// GetLogger returns this controller logger prefilled with basic information.
 	GetLogger() logr.Logger
+
+	// Stop stops the controller and all its watches dynamically.
+	// Note that it will only trigger the stop but will not wait for them all stopped.
+	Stop() error
```

Comment:
Why do we add a `Stop` function here instead of canceling the context that was passed to `Start`?

Reply:
1. Not all of the implementations of `Source` can stop, e.g., `Func` or some custom types.
2. Users have no way to cancel the context passed from internal controller to source, to only stop the single watch.